### DISCUSSION TITLE
Deno types in `eslint`/`tsc` without VSCode extension?

### DISCUSSION THREAD

**Discussion by aaronadamsCA** [denoland/deno#16808] [Q&A]
We have an existing Node.js development environment with many existing Node.js packages/apps in a monorepo. Everything in our repository is written in TypeScript, linted with `eslint`, and type checked with `tsc`.

We've recently added a Remix + Deno Deploy app to our repository. In our dev environment we use `remix watch` and `deno run`; to deploy the app we run `remix build` and deploy `build/index.js` to Deno Deploy.

This required a few workarounds, like patching the Remix `esbuild` config, and running `deno types > deno.d.ts` (and then removing `/// <reference no-default-lib="true" />`) to help ESLint and TypeScript understand things like `Deno.env`. But everything seems to be working; we're successfully developing a Deno app in a Node.js development environment, which is a great first step toward a larger adoption of Deno.

One part I'm not happy with, is that there doesn't seem to be a way to teach ESLint and TypeScript how to understand `deno.land` imports, so that our adoption of Deno can be gradual and seamless. `deno lint` is too limited, and we don't want two separate code quality solutions in one repository; we're also not interested in a VS Code extension, which also wrongly assumes our entire repository and development environment would adopt Deno in one big bang. What we're looking for are tools for gradual adoption in a large existing codebase, which is realistically the only way we'll adopt other parts of Deno besides Deno Deploy.

Are there solutions for any of this? Any guidance would be appreciated, thanks!

**Comment by bartlomieju** [comment]
> we're also not interested in a VS Code extension, which also wrongly assumes our entire repository and development environment would adopt Deno in one big bang

Are you aware that you can enable Deno's VSCode extension only in parts of the repository using `deno.enablePaths` configuration?

> One part I'm not happy with, is that there doesn't seem to be a way to teach ESLint and TypeScript how to understand deno.land imports, so that our adoption of Deno can be gradual and seamless.

You can probably teach `eslint` about them via a plugin, but it's most likely not possible to teach TypeScript how to do it.

**Comment by elycheikhsmail** [comment]
@bartlomieju 
 you said
```
You can probably teach eslint about them via a plugin, but it's most likely not possible to teach TypeScript how to do it.
```
you can tell more about "not possible to teach TypeScript how to donit" or give links to some referance

**Comment by elycheikhsmail** [comment]
some idea

helper library working with deno by default
and working with node (for ide support) using dtn

importmap

some config
